File: src/main/java/sage/entity/Comment.java
package sage.entity;

import java.util.Date;

import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.ManyToOne;

import sage.domain.IdCommons;

@Entity(name = "Comment")
public class Comment {
  private Long id;
  private String content;
  private User author;
  private Date time;
  private Tweet source;

  public Comment() {
  }

  public Comment(String content, User author, Date time, Tweet source) {
    this.content = content;
    this.author = author;
    this.time = time;
    this.source = source;
  }

  @Id
  @GeneratedValue
  public Long getId() {
    return id;
  }

  public void setId(Long id) {
    this.id = id;
  }

  public String getContent() {
    return content;
  }

  public void setContent(String content) {
    this.content = content;
  }

  @ManyToOne(optional = false)
  public User getAuthor() {
    return author;
  }

  public void setAuthor(User author) {
    this.author = author;
  }

  public Date getTime() {
    return time;
  }

  public void setTime(Date time) {
    this.time = time;
  }

  @ManyToOne
  public Tweet getSource() {
    return source;
  }

  public void setSource(Tweet source) {
    this.source = source;
  }

  @Override
  public String toString() {
    return author + ": " + content;
  }

  @Override
  public int hashCode() {
    return IdCommons.hashCode(getId());
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj)
      return true;
    if (obj == null)
      return false;
    if (getClass() != obj.getClass())
      return false;
    
    Comment other = (Comment) obj;
    return IdCommons.equal(getId(), other.getId());
  }
}

File: src/main/kotlin/sage/entity/Comment.kt
package sage.entity

import sage.domain.constraints.CommentConstraints
import javax.persistence.*

@Entity
class Comment(
    @Column(columnDefinition = "TEXT")
    val inputContent: String,

    hyperContent: String,

    @ManyToOne(optional = false) val author: User,

    var sourceType: Short,

    var sourceId: Long,

    val replyUserId: Long?
) : AutoModel() {

  @Column(columnDefinition = "TEXT")
  val content: String = hyperContent

  fun validate() {
    CommentConstraints.check(inputContent)
  }

  companion object : Find<Long, Comment>() {
    val BLOG: Short = 1
    val TWEET: Short = 2

    fun list(sourceType: Short, sourceId: Long): List<Comment> =
        if ((1..2).contains(sourceType)) where().eq("sourceType", sourceType).eq("sourceId", sourceId).findList()
        else emptyList()

    fun count(sourceType: Short, sourceId: Long) =
        if ((1..2).contains(sourceType)) where().eq("sourceType", sourceType).eq("sourceId", sourceId).findRowCount()
        else 0
  }
}